Enable funlen linter

brojeg opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

<!--- Please provide a general summary of your changes in the title above -->

## Pull request type

<!-- Please try to limit your pull request to one types, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: [934](https://github.com/ydb-platform/ydb-go-sdk/issues/934)

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

rekby commented on balancers/config_test.go:
```diff
@@ -9,159 +9,149 @@ import (
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/conn"
 )
 
-func TestFromConfig(t *testing.T) {
```
Here and in other _test files

No need to spit test functions.
The code no need to be reused. Better to disable linter to the function or file with nolint directive:
https://golangci-lint.run/usage/false-positives/#nolint-directive

or by rule:
https://golangci-lint.run/usage/false-positives/#exclude-issues-by-path

rekby reviewed (changes requested):
Hello.

You have done great work for split test functions for separate tests!.
Now the PR broken on other linter rules and can't be merged :(

fix it please.
You can use the commands for install local format utilities:
https://github.com/ydb-platform/ydb-go-sdk/blob/aa0703a9f0a8fc8ed24bcfcfa672e7c2473612a5/.github/workflows/lint.yml#L71-L72

Then start `/.github/scripts/format-all-go-code.sh` from root of the repository for auto-fix the issues by script.

rekby commented:
@brojeg hello, thanks for fixes.

Now the PR denied by lint check.
You can see issues at "Files changed" tab on the PR:
https://github.com/ydb-platform/ydb-go-sdk/pull/1036/files#file-examples-topic-topicreader-topicreader_trace-go-L88
Fix it please.

and the PR failed on tests. I saw reason - it is: https://github.com/ydb-platform/ydb-go-sdk/pull/1036/files#file-examples-topic-topicreader-topicreader_trace-go-L88

It needs to wait until I will fix the issue and merge branches. I do it in few days.

brojeg commented:
@rekby Thank you for the prompt responses and thorough review. I'm actively working on addressing the concerns raised by the funlen linter. As of now, not all files have been fixed. The initial PR was submitted to demonstrate ongoing progress. I've started with refactoring tests, though I realize it wasn't strictly necessary at this stage. Currently, my focus has shifted towards refactoring functions, with the plan to commit changes on a per-file basis. Your feedback has been invaluable in guiding these improvements.

rekby commented:
Hello @brojeg, thanks for the work!

You PR will be long and contain a lot of changes. It mean - it will have a lot of conflicts and difficult to merge, because code in main branch will be changed.

What you think about commit small portion of fixes, when it complete?

But don't commit turn on linter until fix all issue

brojeg commented:
Hello @rekby,

Following your suggestion, I've updated golangci.yml to ignore the funlen linter for now, allowing us to merge smaller, incremental changes. I've pushed a part of the refactored code with this adjustment. Could you please review the PR again? This should help us manage the changes more effectively without overwhelming the merge process.

Thank you for your guidance!

brojeg reviewed (commented):
reviewed the changes

brojeg commented:
> Hello.
> 
> You have done great work for split test functions for separate tests!. Now the PR broken on other linter rules and can't be merged :(
> 
> fix it please. You can use the commands for install local format utilities:
> 
> https://github.com/ydb-platform/ydb-go-sdk/blob/aa0703a9f0a8fc8ed24bcfcfa672e7c2473612a5/.github/workflows/lint.yml#L71-L72
> 
> Then start `/.github/scripts/format-all-go-code.sh` from root of the repository for auto-fix the issues by script.

This issue is resolved.

rekby reviewed (commented):
Hello, sorry for late response.

I have checked part of the PR and will continue in few days.

brojeg commented on balancers/config_test.go:
```diff
@@ -9,159 +9,149 @@ import (
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/conn"
 )
 
-func TestFromConfig(t *testing.T) {
```
reverted the change

brojeg commented on internal/allocator/allocator.go:
```diff
@@ -57,49 +57,57 @@ func New() (v *Allocator) {
-	a.tableQueryAllocator.free()
-	a.tableQueryYqlTextAllocator.free()
-	a.tableQueryIDAllocator.free()
+type Freeable interface {
```
fixed

brojeg commented on internal/backoff/backoff_test.go:
```diff
@@ -41,125 +41,148 @@ func TestDelays(t *testing.T) {
+	runLogBackoffEqualTests(t, backoff, expValues)
+}
+
+func TestLogBackoffWithFullJitterHighCeiling(t *testing.T) {
```
In the initial TestLogBackoff function presents two test cases. These functions differ in their parameters: one uses WithCeiling(6) and the other WithCeiling(3). Additionally, the number of elements in the expValues slice varies between them.

brojeg commented on internal/bind/numeric_args_test.go:
```diff
@@ -326,3 +326,316 @@ SELECT $p0, $p1`,
+	require.ErrorIs(t, err, ErrInconsistentArgs)
+}
+
+func TestNumericArgsBindRewriteQuery_SingleReplacement(t *testing.T) {
```
reverted the change

rekby commented on internal/bind/numeric_args.go:
```diff
@@ -16,9 +16,21 @@ func (m NumericArgs) blockID() blockID {
-func (m NumericArgs) RewriteQuery(sql string, args ...interface{}) (
-	yql string, newArgs []interface{}, err error,
-) {
+func prepareArgs(args []interface{}) ([]interface{}, error) {
```
1. `prepareArgs` is bad name, because I can't understand what do the function by the name. "Prepare" is very common word.

What about `parsePositionalParameters` or `positionalParametersToYDBParams`?

2. `newArgs` contains values with type `*params.Parameter`, in that case will be better to return `[]*params.Parameter` then `[]interface{}`

rekby commented on internal/bind/params.go:
```diff
@@ -21,7 +21,6 @@ var (
 )
 
-//nolint:gocyclo
 func toValue(v interface{}) (_ types.Value, err error) {
```
no need to split the function.

in new style it hase same big type switch, but distributed for few functions with double dependencies (switch case `[]byte, *[]byte, string, *string` must be equal to cases in  `convertBytesAndStrings` it if difficult to read, test, change), it is more difficult, then one big type switch.

rekby commented on internal/bind/params.go:
```diff
@@ -177,63 +244,82 @@ func toYdbParam(name string, value interface{}) (table.ParameterOption, error) {
-				return nil, xerrors.WithStackTrace(err)
-			}
-			params = append(params, param)
+			params, err = appendParam(params, x.Name, x.Value)
```
it is named param too. Why case `case driver.NamedValue` translated to `processNamedValue`, but this case to common `appendParam`?

rekby commented on internal/bind/params.go:
```diff
@@ -177,63 +244,82 @@ func toYdbParam(name string, value interface{}) (table.ParameterOption, error) {
+	}
+}
+
+func appendParam(params []table.ParameterOption, name string, value interface{}) ([]table.ParameterOption, error) {
```
it isn't append parameter only, it additionally convert param to table.ParameterOption (`*params.Parameter` in main branch).

rekby commented on internal/bind/params.go:
```diff
@@ -177,63 +244,82 @@ func toYdbParam(name string, value interface{}) (table.ParameterOption, error) {
 func Params(args ...interface{}) (params []table.ParameterOption, _ error) {
```
Signature of the function was changed in main branch, it needs to rebase the branch and resolve conflicts.

rekby reviewed (commented):
Hello, thanks for the work!
It is better, then yesterday.

It is very big PR, it will be a lot of comments/writers and it will have many conflicts while merge. What about split the PR to few (3-5) small PRs?


For example:
1. Take main branch
2. enable linter (locally)
3. Fix few files
4. Disable the linter
5. Create PR
6. Discussing
7. Merge

Then create next PR same way with respect to previous comments. It will be much faster and simpler for you.

And enable linter in PR on last step only.
